### PR TITLE
Feat/hyper opt: Implement Hyperparameter Tuning

### PR DESCRIPTION
**Pull request by mlederbauer**
Kudos to @kbiniek for the intial experiments on using hyper-opt!

————

This pull request introduces the HyperparameterTuner class and updates the training script to use this new class for hyperparameter tuning. These changes aim to improve the performance of the models used in NMRcraft by optimizing their hyperparameters.

HyperparameterTuner Class
****The new HyperparameterTuner class is located in the [hyperparameter_tune.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) file. This class uses the Hyperopt library to perform hyperparameter tuning. It takes in the name and configuration of a model, an algorithm for hyperparameter tuning (defaulting to tpe.suggest), and the maximum number of evaluations to perform (defaulting to 10).

Here's an example of how to use the HyperparameterTuner class:

```python
tuner = HyperparameterTuner("random_forest", model_config)
best_params = tuner.tune(X_train, y_train, X_test, y_test)
```
In this example, tuner.tune will return the best hyperparameters found for the "random_forest" model.

**Updated Training Script**
The training script has been updated to use the HyperparameterTuner class for hyperparameter tuning. This means that, instead of using default hyperparameters, the script will now search for the best hyperparameters for each model.

There are a few TO-DOs I added that will be added to address future issues/features! Foremostly, the data loader to be added to nmrcraft/data.

As always, please review these changes and let me know if you have any questions or feedback.

### CODE REVIEW

**Comment by mlederbauer** [comment]
Also please note we have a `model_config` now where we can store the model names and all relevant hyperparameters in a dictionary in `nmrcrfaft/models/model_configs.py`. I’ll expand more on them and for example add them to `nmrcraft/models/models.py` once merged !

File: nmrcraft/training/hyperparameter_tune.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            X_test: The testing data.
+            y_test: The testing labels.
+
+        Returns:
+            dict: The loss and status of the objective function.
+        """
+        model = load_model(self.model_name, **{**params, **self.model_config["model_params"]})
+        model.fit(X_train, y_train)
+        y_pred = model.predict(X_test)
+        score = accuracy_score(y_test, y_pred)
```
Comment (kbiniek) [inline nmrcraft/training/hyperparameter_tune.py:42]: we could also include different ways of scoring, for example if i remember correctly they used `score = cross_val_score(model, X_train, y_train, cv=4).mean()` in the tutorial

File: scripts/training/train_metal.py [hunk 2]
```diff
@@ -2,27 +2,28 @@
...
 from nmrcraft.utils.set_seed import set_seed
 
 set_seed()
 
 
 def main(dataset_size, target, model_name):
-    """Train a model with MLflow tracking."""
-    mlflow.set_experiment("Model_Classification_Experiment")
+    # TODO: better experiment naming
+    mlflow.set_experiment("Ceci_nest_pas_un_experiment")
```
Comment (kbiniek) [inline scripts/training/train_metal.py:21]: Loving the title

File: nmrcraft/models/hyperparameter_tune.py
Comment (kbiniek) [inline nmrcraft/models/hyperparameter_tune.py:1]: Do we want to drop the idea of gridsearch or maybe implement both gridsearch and hyper-opt?

**Review by kbiniek — APPROVED** [review]
Great stuff!! Thank you :)

File: scripts/training/train_metal.py (hunk 2 above)
Comment (mlederbauer) [reply scripts/training/train_metal.py:21]: Hahaha thanks we can def work on that tho

File: nmrcraft/training/hyperparameter_tune.py (hunk 1 above)
Comment (mlederbauer) [reply nmrcraft/training/hyperparameter_tune.py:42]: Great idea! Depending on which metric is the most important to us!

File: nmrcraft/models/hyperparameter_tune.py
Comment (mlederbauer) [reply nmrcraft/models/hyperparameter_tune.py:1]: Puh if I got it correctly hyperopt might even be a better method, no? Using Bayesian Optimization to obtain optimal hyperparfameters. We can of course do this is hyperopt turns out to be not too suitable for us!

**Comment by mlederbauer** [comment]
Thank you @kbiniek !

File: nmrcraft/models/hyperparameter_tune.py
Comment (kbiniek) [reply nmrcraft/models/hyperparameter_tune.py:1]: Ah no then lets not! Nevermind my question, didnt look into the details mb
